bittensor/utils/__init__.py

```diff
@@ -671,41 +676,47 @@ def update_curr_block(block_number: int, block_bytes: bytes, diff: int, lock: mu
                 # Set new block events for each solver
                 for w in solvers:
                     w.newBlockEvent.set()
-                    
-            # Get times for each solver
-            time_total = 0
+            
             num_time = 0
+            # Get times for each solver
             for _ in solvers:
                 try:
-                    time_ = time_queue.get_nowait()
-                    time_total += time_
+                    _ = time_queue.get_nowait()
                     num_time += 1
 
                 except Empty:
                     break
             
+            time_since = time.time() - start_time
             if num_time > 0:
-                time_avg = time_total / num_time
-                itrs_per_sec = TPB*update_interval*num_processes / time_avg
-                time_since = time.time() - start_time
+                time_since_last = time.time() - time_last
+                # create EWMA of the hash_rate to make measure more robust
+                hash_rate_ = ((num_time * TPB * update_interval) / time_since_last)
+                hash_rates.append(hash_rate_)
+                hash_rates.pop(0) # remove the 0th data point
+                hash_rate = sum([hash_rates[i]*weights[i] for i in range(n)])/(sum(weights))
+
+                # update time last to now
+                time_last = time.time()
+                
                 
             message = f"""Solving 
                 time spent: {time_since}
                 Difficulty: [bold white]{millify(difficulty)}[/bold white]
-                Iters: [bold white]{get_human_readable(int(itrs_per_sec), 'H')}/s[/bold white]
+                Iters: [bold white]{get_human_readable(int(hash_rate), 'H')}/s[/bold white]
                 Block: [bold white]{block_number}[/bold white]
                 Block_hash: [bold white]{block_hash.encode('utf-8')}[/bold white]"""
             status.update(message.replace(" ", ""))
         
-        # exited while, found_solution contains the nonce or wallet is registered
-        if solution is not None:
-            stopEvent.set() # stop all other processes
-            status.stop()
+        # exited while, solution contains the nonce or wallet is registered
+        stopEvent.set() # stop all other processes
+        status.stop()
 
-            return solution
+        for w in solvers: # terminate and wait for all solvers
+            w.terminate()
+            w.join()
```

Comment:
given that you are repeating this little piece of code.. Would it have sense to extract into a function and use it?